Use FeatureIdIndexOption in RecreateFactsetJob

charleso opened the pull request:
Isolating this change to continue discussion about missing featureId behaviour.

https://github.com/ambiata/ivory/pull/626#discussion_r25660540

charleso commented on ivory-storage/src/main/scala/com/ambiata/ivory/storage/repository/RecreateFactsetJob.scala:
```diff
@@ -109,14 +110,17 @@ class RecreateFactsetMapper extends Mapper[NullWritable, BytesWritable, BytesWri
-    val k = featureIdLookup.ids.get(fact.featureId.toString).toInt
+    val featureId = FeatureIdIndexOption.lookup(fact, featureIdLookup)
+    if (featureId.isEmpty)
+      Crash.error(Crash.Invariant, s"Unknown feature '${fact.featureId}' for entity '${fact.entity}'")
```
Yeah, so this did used to crash, but sounds like it shouldn't and just use the string instead?

markhibberd replied:
I don't think this matches the current behaviour, unless I am misreading, and I don't think it is correct, it should be copying across all the facts.

charleso commented:
I have added a fix that uses `FeatureIdIndex(0)` and modified the test. It's currently failing because rename a `Factset` with 100% unknown features will result in 0 reducers. What is the expected behaviour in this case? /cc @raronson

markhibberd commented on ivory-storage/src/main/scala/com/ambiata/ivory/storage/repository/RecreateFactsetJob.scala:
```diff
@@ -109,9 +110,10 @@ class RecreateFactsetMapper extends Mapper[NullWritable, BytesWritable, BytesWri
 
-    FactsetWritable.set(fact, kout, k)
+    // If we don't know about this feature any more just default to 0 - don't fail
+    FactsetWritable.set(fact, kout, if (featureId.isDefined) featureId.get else FeatureIdIndex(0))
```
Is 0 the starting index by default? If so, could we pick something larger, maybe `max(featureId) + 1` if it is possible. The main goal here is to return to a close as possible to perfect state - if there are features not currently in the dictionary I wouldn't wanting messing up compression of attribute 0.

markhibberd commented:
Besides the default index, I am good with this. In the 0 case, it should fall back to some dumb size calculation (1 reducer for ever 1GB of data maybe)

charleso commented on ivory-storage/src/main/scala/com/ambiata/ivory/storage/task/FactsPartitioner.scala:
```diff
@@ -38,9 +38,13 @@ trait BaseFactsPartitioner[A] extends Partitioner[A, BytesWritable] with Configu
+  def getPartition(k: A, v: BytesWritable, partitions: Int): Int = {
+    val offset = lookup.reducers.get(getFeatureId(k).int)
+    // If we don't know about this feature partition across all reducers
+    if (offset == null) getHash(k) % partitions
```
@markhibberd This is related to your comment about `max + 1` this morning. I don't think this is right. Should it be `featureId % partitions` instead, and does that mean we might get some nasty skew?

nhibberd commented:
@charleso bump

markhibberd commented:
@charleso Bump. I don't know context on this - if it is just tidy up, should we just close given how stale it is - or was this to address some issue?

charleso commented:
@markhibberd Yeah, basically tidy up, no issue.
